File: src/main/java/org/mercurialftc/mercurialftc/scheduler/commands/Command.java
package org.mercurialftc.mercurialftc.scheduler.commands;

import org.mercurialftc.mercurialftc.scheduler.OpModeEX;
import org.mercurialftc.mercurialftc.scheduler.Scheduler;
import org.mercurialftc.mercurialftc.scheduler.subsystems.SubsystemInterface;

import java.util.Set;

public interface Command {
	/**
	 * Gets run once when a command is first scheduled
	 */
	void initialise();

	/**
	 * Gets run once per loop until {@link #finished()}
	 */
	void execute();

	/**
	 * Gets run once when {@link #finished()} or when interrupted
	 */
	void end(boolean interrupted);

	/**
	 * the supplier for the natural end condition of the command
	 *
	 * @return true when the command should finish
	 */
	boolean finished();

	/**
	 * @return the set of subsystems required by this command
	 */
	Set<SubsystemInterface> getRequiredSubsystems();

	/**
	 * @return the set of OpMode run states during which this command is allowed to run
	 */
	Set<OpModeEX.OpModeEXRunStates> getRunStates();

	/**
	 * @return if this command is allowed to be interrupted by others
	 */
	default boolean interruptible() {
		return true;
	}

	/**
	 * schedule the command with the scheduler
	 */
	default void queue() {
		if (!Scheduler.getSchedulerInstance().isScheduled(this)) Scheduler.getSchedulerInstance().scheduleCommand(this);
	}
}

File: Mercurial/src/main/kotlin/dev/frozenmilk/mercurial/commands/Command.kt
package dev.frozenmilk.mercurial.commands

import dev.frozenmilk.mercurial.Mercurial
import dev.frozenmilk.mercurial.RunStates
import dev.frozenmilk.mercurial.subsystems.Subsystem

interface Command {
	/**
	 * Gets run once when a command is first scheduled
	 */
	fun initialise()

	/**
	 * Gets run once per loop until [.finished]
	 */
	fun execute()

	/**
	 * Gets run once when [.finished] or when interrupted
	 */
	fun end(interrupted: Boolean)

	/**
	 * the supplier for the natural end condition of the command
	 *
	 * @return true when the command should finish
	 */
	fun finished(): Boolean

	/**
	 * @return the set of subsystems required by this command
	 */
	val requiredSubsystems: Set<Subsystem>

	/**
	 * @return the set of OpMode run states during which this command is allowed to run
	 */
	val runStates: Set<RunStates>

	/**
	 * @return if this command is allowed to be interrupted by others
	 */
	val interruptible: Boolean
		get() = true

	/**
	 * schedule the command with the scheduler
	 */
	fun schedule() {
		if (!Mercurial.isScheduled(this)) Mercurial.scheduleCommand(this)
	}

	fun toLambda(): LambdaCommand = LambdaCommand.from(this)
}
